Implement New Lock Reservation scheme on Power

IBMJimmyk opened the pull request:
This commit is for a new lock reservation scheme that is currently only active on power.

The new lock reservation scheme adds a new Learning phase before a lock is reserved. During the Learning phase a lock reservation cancellation is not required if a second thread wishes to lock an object. If the Learning phase is successful, the object can be reserved.

If two threads attempt to lock the same object, then it can never be reserved again. Under the old system, lock reservation could be attempted again on the same object even after a reservation on the object was cancelled.

A system to bypass the new Learning state is added via new initial lockwords. Data is collected on a per class basis to try and predict if locked objects of that class will eventually go to Flat or Reserved. Based on that data, objects might go straight to Flat or Reserved when they are locked for the first time.

To activate the new lock reservation scheme use the -XX:+GlobalLockReservation command line option. Without the option, the lock reservation system used will be very similar to the original system.

Minor changes to x86 codegen were made to accomadate the lockword format change required to implement the new lock reservation scheme.

This also removes much of the unused PPC Monitor Snippet code.

More details can be found in the documentation at:
doc/compiler/ObjectLockword.md

Implements: https://github.com/eclipse/openj9/issues/8045

gita-omr commented:
@zl-wang please take a look if you have time.

gita-omr commented:
@pshipton @andrewcraik @gacholio we would like to get this into 0.18 as an experimental feature under an option (not changing the current default behaviour) but please let us know if it's too short of a notice to verify that.

gacholio commented:
I'm on vaction in a few days, so I doubt I'll be looking at this.

andrewcraik commented on doc/compiler/ObjectLockword.md:
```diff
@@ -0,0 +1,386 @@
+
+The TID field takes advantage of how the lowest 8 bits in a Thread ID are always 0 so they don't need to be stored.
+
+The RC or Recursion Count field is primarily used to track nested locking. One key point is the RC field is slightly different for reserved locks and unreserved locks. For reserved locks the RC field is equal to the nesting depth of the locks. This means that if the object was locked 3 times before being unlocked then the RC field will have a value of 3. If the object is not locked at all, the RC ...
```
Would be helpful to note the behaviour here on overflow of the counter...

IBMJimmyk replied:
I will add that in.

andrewcraik commented on doc/compiler/ObjectLockword.md:
```diff
@@ -0,0 +1,386 @@
+
+The TID field is largely handled the same way as before. It stores the TID (minus the lower 8 bits which are always 0) of the thread that currently owns the lock or the thread that reserved the lock. It can now also hold the TID of a thread that is trying to reserve the lock. This is connected to the new Learning state.
+
+The RC or Recursion Count field is primarily used to track nested locking. Reserved locks and Flat locks are treated the same way as before. Locks in the Learning state handle RC in the same was as lock in the Reserved state. If a lock has been locked exactly 2 times with no unlocking actions in between and the lock is currently in the Learning state, then the RC will hold a value of 2.
```
Behaviour on RC overflow in Learning would be helpful to clarify

IBMJimmyk replied:
I will add that in as well.

andrewcraik commented on doc/compiler/ObjectLockword.md:
```diff
@@ -0,0 +1,386 @@
+
+The RC or Recursion Count field is primarily used to track nested locking. Reserved locks and Flat locks are treated the same way as before. Locks in the Learning state handle RC in the same was as lock in the Reserved state. If a lock has been locked exactly 2 times with no unlocking actions in between and the lock is currently in the Learning state, then the RC will hold a value of 2.
+
+The LC or Learning Count is new. This field only appears for locks in the Learning state. It gets incremented whenever a lock is locked by the same thread. After it reaches a certain threshold the lock will transition from the Learning state to the Reserved state.
```
Behaviour on LC overflow in Learning would be helpful to clarify

IBMJimmyk replied:
Same as other two comments, I'll add them in.

gacholio commented:
I have not looked the code, but I have a question - is this handling runtime compressed refs? If you have added no ifdefs for compressed, then I think we're OK.

IBMJimmyk commented:
My code changes do not use any ifdefs for handing compressedrefs.

dmitripivkine commented on runtime/gc_base/ObjectAccessBarrier.cpp:
```diff
@@ -1552,10 +1584,47 @@ MM_ObjectAccessBarrier::cloneIndexableObject(J9VMThread *vmThread, J9IndexableOb
 	lockwordAddress = getLockwordAddress(vmThread, (J9Object*) destObject);
 	if (NULL != lockwordAddress) {
-		J9_STORE_LOCKWORD(vmThread, lockwordAddress, OBJECT_HEADER_LOCK_RESERVED);
+		U_32 lwValue = 0;
```
Is this code copy/paste from copyObjectFields()? If it is identical please put it in the function

IBMJimmyk replied:
Yes, the code for determining the initial lockword is very similar and occurs in a few places. I will try to combine them into one place.

dmitripivkine commented on runtime/gc_base/ObjectAccessBarrier.cpp:
```diff
@@ -1517,11 +1517,43 @@ MM_ObjectAccessBarrier::copyObjectFields(J9VMThread *vmThread, J9Class *objectCl
+
+			if (vmThread->javaVM->enableGlobalLockReservation) {
+				/* This path is taken when Global Lock Reservation is enabled. */
+				U_32 lockReservationHistory = (U_32)objectClass->lockReservationHistory;
```
I believe for `lockReservationHistory` no need to be `U_32`

IBMJimmyk replied:
I've now changed that to UDATA.

dmitripivkine commented:
@amicic FYI

gita-omr commented:
Attention reviewers: please finish your reviews.

dmitripivkine commented on runtime/oti/ObjectMonitor.hpp:
```diff
@@ -473,6 +482,113 @@ class VM_ObjectMonitor
+			 * Start as New-PreLearning -> Initial lockword: OBJECT_HEADER_LOCK_LEARNING
+			 * Start as Flat-Unlocked   -> Initial lockword: 0
+			 */
+			if (reservedCount >= reservedAbsoluteThreshold && (reservedCount > (cancelCount * minimumReservedRatio))) {
```
please add extra brackets `(reservedCount >= reservedAbsoluteThreshold)`

IBMJimmyk replied:
Okay, I'll add that.

dmitripivkine commented on runtime/oti/ObjectMonitor.hpp:
```diff
@@ -473,6 +482,113 @@ class VM_ObjectMonitor
+			 */
+			if (reservedCount >= reservedAbsoluteThreshold && (reservedCount > (cancelCount * minimumReservedRatio))) {
+				initialLockword = OBJECT_HEADER_LOCK_RESERVED;
+			} else if (cancelCount < cancelAbsoluteThreshold || (reservedCount > (cancelCount * minimumLearningRatio))) {
```
please add extra brackets `(cancelCount < cancelAbsoluteThreshold)`

IBMJimmyk replied:
I'll add this as well.

dmitripivkine commented on runtime/gc_include/ObjectAllocationAPI.hpp:
```diff
@@ -29,6 +29,7 @@
 #include "j9consts.h"
 #include "omrgcconsts.h"
 #include "AtomicSupport.hpp"
+#include "ObjectMonitor.hpp"
```
Would you please the same explicit include to `ObjectAccessBarrier.cpp` and `mgcalloc.cpp`?

IBMJimmyk replied:
I can do this. Is there a particular order the includes should be in? The existing includes aren't quite alphabetical.

dmitripivkine replied:
Well theoretically  includes should be sorted in alphabetical order (in two groups for `*.h` and `*.hpp` files). However in practice it is not used everywhere. `mgcalloc.cpp` has includes sorted so you can include it in proper place. `ObjectAccessBarrier.cpp` has includes not sorted so you can put it somewhere

IBMJimmyk replied:
From trying to do this I found a problem. It looks like putting `getInitialLockword` inside `ObjectMonitor.hpp` causes a circular reference when the `ObjectMonitor.hpp` file is included in `ObjectAllocationAPI.hpp`. `ObjectAllocationAPI.hpp` uses `getInitialLockword`.

Include loop:
ObjectMonitor.hpp -> VMHelpers.hpp -> ObjectAllocationAPI.hpp -> ObjectMonitor.hpp

So I think I'll need to move `getInitialLockword` somewhere else or break the circle in a different way.

IBMJimmyk replied:
I talked to @amicic and it looks like the include loop can be broken by removing the `VMHelpers.hpp` include inside `ObjectMonitor.hpp`. For what we could tell it wasn't needed. Builds on x86, Power and z were still successful even after removing the include.

dmitripivkine commented on runtime/gc_modron_startup/mgcalloc.cpp:
```diff
@@ -114,9 +114,13 @@ J9AllocateObjectNoGC(J9VMThread *vmThread, J9Class *clazz, uintptr_t allocateFla
-					J9_STORE_LOCKWORD(vmThread, lockEA, OBJECT_HEADER_LOCK_RESERVED);
+				if (LN_HAS_LOCKWORD(vmThread, objectPtr)) {
+					j9objectmonitor_t initialLockword = VM_ObjectMonitor::getInitialLockword(vmThread->javaVM, clazz);
+
```
Would you please remove this empty line?

IBMJimmyk replied:
I'll remove it.

dmitripivkine commented on runtime/gc_modron_startup/mgcalloc.cpp:
```diff
@@ -381,9 +385,13 @@ J9AllocateObject(J9VMThread *vmThread, J9Class *clazz, uintptr_t allocateFlags)
-				J9_STORE_LOCKWORD(vmThread, lockEA, OBJECT_HEADER_LOCK_RESERVED);
+			if (LN_HAS_LOCKWORD(vmThread, objectPtr)) {
+				j9objectmonitor_t initialLockword = VM_ObjectMonitor::getInitialLockword(vmThread->javaVM, clazz);
+
```
Would you please remove this empty line?

IBMJimmyk replied:
I'll remove it.

dmitripivkine commented on runtime/gc_base/ObjectAccessBarrier.cpp:
```diff
@@ -1517,11 +1517,9 @@ MM_ObjectAccessBarrier::copyObjectFields(J9VMThread *vmThread, J9Class *objectCl
-				J9_STORE_LOCKWORD(vmThread, lockwordAddress, 0);
-			}
+			j9objectmonitor_t lwValue = VM_ObjectMonitor::getInitialLockword(vmThread->javaVM, objectClass);
+
```
Would you please remove this empty line?

IBMJimmyk replied:
I'll remove it.

dmitripivkine commented on runtime/gc_base/ObjectAccessBarrier.cpp:
```diff
@@ -1552,10 +1550,13 @@ MM_ObjectAccessBarrier::cloneIndexableObject(J9VMThread *vmThread, J9IndexableOb
 	if (NULL != lockwordAddress) {
-		J9_STORE_LOCKWORD(vmThread, lockwordAddress, OBJECT_HEADER_LOCK_RESERVED);
+		j9objectmonitor_t lwValue = VM_ObjectMonitor::getInitialLockword(vmThread->javaVM, objectClass);
+
```
Would you please remove this empty line?

IBMJimmyk replied:
I'll remove it.

dmitripivkine commented on runtime/gc_include/ObjectAllocationAPI.hpp:
```diff
@@ -334,19 +332,25 @@ class MM_ObjectAllocationAPI
+
+		if (initializeSlots) {
+			if (LN_HAS_LOCKWORD(currentThread, instance)) {
+				j9objectmonitor_t initialLockword = VM_ObjectMonitor::getInitialLockword(currentThread->javaVM, clazz);
```
Is this code breaks run time Compressed/Full support added to this function already?

IBMJimmyk replied:
No, it should be compatible. The initial lockword value can only be one of 3 raw values: 0 (never reserve), 4 (go to Reserved state on first lock) or 8 (go to Learning state on first lock). It never sets the TID field so it works for both 32 bit and 64 bit lockwords.

A few lines later the macro `J9_STORE_LOCKWORD` may be used to set the new lockword. `J9_STORE_LOCKWORD` is defined in `j9accessbarrier.h` and uses the `J9VMTHREAD_COMPRESS_OBJECT_REFERENCES` macro to determine if compressed refs are used or not. The width of the store is adjusted accordingly.

dmitripivkine replied:
ok, I see an implementation of `J9_STORE_LOCKWORD`... Logically looks correct however would introduce check of `if (J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(vmThread))` for second time

dmitripivkine replied:
also there is difference `J9_STORE_LOCKWORD` definition uses  extra `volatile` declaration. I am not sure can it have performance impact in this case.

IBMJimmyk replied:
Do you think it would be better to change this to a non-volatile store? There's also changes in `objectMonitorExit` in `monhelpers.c` where I changed the code to use `J9_STORE_LOCKWORD` that may also need to be changed to use non-volatile stores.

zl-wang replied:
I don't think it is necessary to use volatile store here in sequence of Object initialization, since there will be memory barrier later anyway.

IBMJimmyk commented:
Previously requested changes have been made except for the one regarding the volatile keyword since I'm not sure if it needs to be changed.

Also, @gacholio did you want to take a look at this PR?

gita-omr commented:
We would like finish this work soon. @pshipton what do you think about running existing tests under this option? I think we also need to document it externally.

gita-omr commented on doc/compiler/ObjectLockword.md:
```diff
@@ -0,0 +1,386 @@
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
```
Nice document! Minor suggestions:

1) Would be nice to have a table of content, at least for the major sections
2) Link to the main issue
3) Maybe a diagram?
4) Would be nice to explain why an object can never return to reserved state once it was canceled.

gita-omr commented:
@andrewcraik @gacholio @dmitripivkine @zl-wang:  please finish your reviews this week.

pshipton commented:
> what do you think about running existing tests under this option

Yes, good plan. Don't forget to at least run all 4 suites, sanity.functional, extended.functional, sanity.system extended.system. Grinders allow you to add extra options, although I'd lean towards creating a commit that enables the option by default and use that commit to run all the testing.

gita-omr commented:
> 
> 
> > what do you think about running existing tests under this option
> 
> Yes, good plan. Don't forget to at least run all 4 suites, sanity.functional, extended.functional, sanity.system extended.system. Grinders allow you to add extra options, although I'd lean towards creating a commit that enables the option by default and use that commit to run all the testing.

@IBMJimmyk did that. I was thinking of adding an extra mode to our existing tests and run them regularly, eg. as a part of acceptance build. Do we have a precedent for that?

dmitripivkine commented:
> > > what do you think about running existing tests under this option
> > 
> > 
> > Yes, good plan. Don't forget to at least run all 4 suites, sanity.functional, extended.functional, sanity.system extended.system. Grinders allow you to add extra options, although I'd lean towards creating a commit that enables the option by default and use that commit to run all the testing.
> 
> @IBMJimmyk did that. I was thinking of adding an extra mode to our existing tests and run them regularly, eg. as a part of acceptance build. Do we have a precedent for that?

Well if we need prolonged soaking period we can create special spec(s) for this feature. We did it for Concurrent Scavenger successfully. The only concern for this approach do we have enough resources rot run one or more new specs

IBMJimmyk commented:
Some of the check were failing so I've rebased my latest changes on the newest master. I also resolved the conflicts from the changes to accessing `TR::Compiler->target`.

pshipton commented:
> do we have enough resources to run one (or more) new specs

I expect we do for Power, although probably not the special.system testing, but it could run on the weekend.

gita-omr commented:
Definitely, just running on weekends would be great. Are there any instructions how to set it up @smlambert ?

gacholio commented on runtime/vm/jvminit.c:
```diff
@@ -2253,6 +2256,45 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
+			argIndex = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXNOGLOBALLOCKRESERVATION, NULL);
+			argIndex2 = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXGLOBALLOCKRESERVATION, NULL);
+
+			if (argIndex2 >= 0 && argIndex2 > argIndex) {
```
Please bracket the expressions.

IBMJimmyk replied:
Okay, I'll change that.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ObjectMonitor_V2.java:
```diff
@@ -0,0 +1,322 @@
+import com.ibm.j9ddr.vm29.types.IDATA;
+import com.ibm.j9ddr.vm29.types.UDATA;
+
+class ObjectMonitor_V2 extends ObjectMonitor
```
I would prefer not to do that, `ObjectMonitor_V1` should not change going forward; keeping them separate supports that.

gita-omr commented:
@keithc-ca could you approve all the DDR related changes?

gita-omr commented:
@IBMJimmyk :  just wanted to confirm: the only changes in J9PPCSnipet.cpp/hpp is removal of unused code? Then I guess we only need to review J9TreeEvaluator.cpp for Power changes @zl-wang

IBMJimmyk commented:
`J9PPCSnippet.cpp` changes were just the removal of unused code.
`J9PPCSnippet.hpp` changes were mostly just the removal of unused code. I also moved some of the defines from `J9PPCSnippet.hpp` to `J9TreeEvaluator.cpp` since they were no longer used in any snippets and only used in `J9TreeEvaluator.cpp`.

zl-wang commented on runtime/oti/ObjectMonitor.hpp:
```diff
@@ -473,6 +481,113 @@ class VM_ObjectMonitor
+				initialLockword = OBJECT_HEADER_LOCK_RESERVED;
+			} else if ((cancelCount < cancelAbsoluteThreshold) || (reservedCount > (cancelCount * minimumLearningRatio))) {
+				initialLockword = OBJECT_HEADER_LOCK_LEARNING;
+			}
```
I felt that always accessing the two sub-fields together increases the chance of value drifting (i.e.  multi-threads trash each other's updates).  By and large, each routine mostly increments (modifies) one sub-field, except under overflow conditions (very very rare ... we can ignore). Accessing one sub-field would avoid potential trashing mostly.

zl-wang replied:
One minor side-effect of accessing the whole UDATA:  the data field locations are different in memory image on Big-Endian and Little-Endian platforms.  Likely it doesn't matter. Consumability?

IBMJimmyk replied:
I'm making a change to separate `reservedCounter` and `cancelCounter` into two `U_16` values in `J9Class`. This means that they are now updated independently of each other. So incrementing one counter will not touch the other. However, this will also mean when handling overflow at the point where the counters are shifted right by 1 (to divide them in half) it will be possible to read in between updating the two counters.

gacholio replied:
Note there is no implicit alignment in J9Class, so you can't have U16/U16/UDATA. Either the U16s will have to take the place of an existing U32, or U32 padding will need to be added on 64-bit platforms.

Pretty much all of our structures assume that their fields are aligned to their type - this is especially important for fields that we use atomics on.

IBMJimmyk replied:
I was planning to do something like this:
```
union {
	UDATA padding; // This is used to preserve alignment under 64 bit.
	struct {
		U_16 reservedCounter;
		U_16 cancelCounter;
	} counters;
} lockReservationHistory;
```
The UDATA should preserve the alignment while adding in the two 16bit counters. Will something like that work?

zl-wang replied:
I don't see the need to go that complicated.  for 64bit, go with two U_32;  for 32bit, go with two U_16.  Either can be used to count to top of 0xFFFF.  Any drawback of this?

gacholio replied:
Personally, I prefer the explicit ifdef padding, so that's it's obvious (we don't want to start accumulating lots of these unions that waste padding space).

gacholio replied:
Having the fields be different types will be a pain for both the C code and the JIT code (you'll be making assumptions about the field size under ifdef).

IBMJimmyk replied:
I decided to go with putting in an explicit ifdef to handle padding. Changes are in the new commit I just added.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5870,128 +5906,124 @@ TR::Register *J9::Power::TreeEvaluator::VMmonexitEvaluator(TR::Node *node, TR::C
+         {
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, tempReg, monitorReg, 0, ~LOCK_NON_PRIMITIVE_EXIT_IGNORE_MASK);
+         }
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::Op_cmpl, node, condReg, tempReg, metaReg);
```
Given the pending integration for a unified driver from @gacholio ,  the choice of certain opCode(s) had better to be compile-time dynamic, instead of Op_cmpl way of being statically decided by #if.  I would strongly suggest you go over the whole code you modified at least to catch all the cases.

IBMJimmyk replied:
I forgot to mention earlier but I've changed this. Instead of using variations of `Op_cmp`, there is a runtime check for compressed lockwords and whether it is 32bit or 64bit to select the cmp instruction to use. I only made the changes to the code sections I modified so there are still some uses of `Op_cmp` in lockword related code paths (eg. old reservation path) that I didn't update.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -76,6 +76,19 @@
+#define LOCK_RES_OWNING_COMPLEMENT                        (OBJECT_HEADER_LOCK_RECURSION_MASK | OBJECT_HEADER_LOCK_FLC)
+
+#define LOCK_NON_PRIMITIVE_ENTER_IGNORE_MASK              ((OBJECT_HEADER_LOCK_RECURSION_MASK & ~OBJECT_HEADER_LOCK_LAST_RECURSION_BIT) | OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_FLC)
+#define LOCK_NON_PRIMITIVE_EXIT_IGNORE_MASK               (OBJECT_HEADER_LOCK_RECURSION_MASK | OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_FLC)
```
Masks used in rotate-masking instructions have to be a single-range circularly.  I assumed you already made sure of this.

IBMJimmyk replied:
Most of the masks only deal with the lowest 8 bits. The exceptions are the ones that need to look at the TID stored in the lockword. I've made sure that the masks used by instructions like `rlwinm` are compatible with the instruction.

IBMJimmyk replied:
I double checked this and I actually made a mistake earlier. Those masks will not work with `rlwinm` due to the position of the Learning bit. So I will have to remove that path.

zl-wang commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5870,128 +5906,124 @@ TR::Register *J9::Power::TreeEvaluator::VMmonexitEvaluator(TR::Node *node, TR::C
+       * callLabel:
+       *    bl    jitMonitorExit
+       *    b     callReturnLabel
+       */
```
Performance implications of code layout ... have you looked into it?  i.e. more frequently travelled code should be arranged to fall through, avoiding taken branches which hardware will take anywhere 2-4 cycles to turn around, even assuming that it is perfectly predicted.

IBMJimmyk replied:
I've tried to minimize the number of taken branches on the fast path. Most of the taken branches are done when a check says to go to the VM to handle the current state. There's also a branch that selects between the nested and non-nested case that can't be avoided. It currently falls through to the non-nested case since that's what the original code did.

keithc-ca commented on runtime/ddr/algorithm_versions.c:
```diff
@@ -41,6 +41,7 @@
 #define ALG_VM_VTABLE_VERSION 1
 #define ALG_VM_ITABLE_VERSION 1
 #define ALG_VM_BYTECODE_VERSION 1
+#define OBJECT_MONITOR_VERSION 1
```
This should probably be named `ALG_ OBJECT_MONITOR_VERSION` given the established pattern here.

IBMJimmyk replied:
I used `OBJECT_MONITOR_VERSION` because it was already called `OBJECT_MONITOR_VERSION` in a few places like in `AlgorithmVersion.java` and `ObjectMonitor.java`. Should I change all of them to `ALG_OBJECT_MONITOR_VERSION`?

gacholio replied:
For sanity, I'd make the constant 2 instead of 1, just so it matches the class name.

keithc-ca replied:
That makes sense, but we have a long history of doing something different. We have the class `ObjectMonitor_V1` that applies to old core files. If we want to reduce confusion, I suggest we rename that class `ObjectMonitor_V0` and use `ObjectMonitor_V1` for the new one. We could apply the same renaming pattern to all the version-specific classes (in a separate piece of work).

I think we should use `ALG_OBJECT_MONITOR_VERSION` for the new version constant being introduced: the references to `OBJECT_MONITOR_VERSION` should be updated. Neither name will be in old core files so the old path will be taken.

keithc-ca commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -2827,10 +2827,19 @@ typedef struct J9Object {
+#define OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT  0x10
 #define OBJECT_HEADER_LOCK_LAST_RECURSION_BIT  0x80
-#define OBJECT_HEADER_LOCK_RECURSION_MASK  0xF8
-#define OBJECT_HEADER_LOCK_RECURSION_OFFSET  3
```
This old value must be used within `ObjectMonitor_V1` otherwise old core files will be misinterpreted. Perhaps we should keep the old macros and use new in `ObjectMonitor_V2` and elsewhere.

gacholio replied:
Would it make sense to just use newly-named constants and leave the old ones alone (documented so they don't get deleted later!)?

IBMJimmyk replied:
There are 3 defines that need new values: `OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT`, `OBJECT_HEADER_LOCK_RECURSION_MASK` and `OBJECT_HEADER_LOCK_RECURSION_OFFSET`.

I understand the justification for leaving in `OBJECT_HEADER_LOCK_RECURSION_OFFSET` for use in `ObjectMonitor_V1`. But, I don't really understand why old versions of the other two are still needed if only the new values are used. Is there a particular reason you feel like it might be useful  to keep them around? If not, it seems like it is just adding more complications without any real benefit.

keithc-ca replied:
Any constants used by `ObjectMonitor_V1` must not change else we will be unable to properly interpret old core files (where `ALG_OBJECT_MONITOR_VERSION` is 0). We can either add new constants to be used for the new implementation (with comments warning against new uses of the old constants) or inline the old constant values into `ObjectMonitor_V1`. Graham suggested the former plan and I agree it is the better approach.

IBMJimmyk replied:
I understand that `ObjectMonitor_V1` needs the older values. However, it only looks like `OBJECT_HEADER_LOCK_RECURSION_OFFSET` is used inside `ObjectMonitor_V1`. `OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT` and `OBJECT_HEADER_LOCK_RECURSION_MASK` are not used in `ObjectMonitor_V1`.

keithc-ca replied:
Ok, `OBJECT_HEADER_LOCK_RECURSION_OFFSET` is the only constant used by `ObjectMonitor_V1` that's changing now. If other used constants change in the future, that will break support for old core files. I suggest we need a comment that warns about changing the constants that must not change without fixing `ObjectMonitor_V1` accordingly.

I think the constants for the new implementation should use a different prefix: I suggest `OBJECT_HEADER_LOCK_V2_`.

IBMJimmyk replied:
Okay then I will make the following changes:
New `OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT` -> `OBJECT_HEADER_LOCK_V2_FIRST_RECURSION_BIT`
New `OBJECT_HEADER_LOCK_RECURSION_MASK` -> `OBJECT_HEADER_LOCK_V2_RECURSION_MASK`
New `OBJECT_HEADER_LOCK_RECURSION_OFFSET` -> `OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET`

I will also add a comment regarding the connection to `ObjectMonitor` java files.

IBMJimmyk replied:
I've added the changes in the latest commit.

keithc-ca commented on runtime/vm/jvminit.c:
```diff
@@ -2253,6 +2256,45 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
+				/* Global Lock Reservation is currently only supported on Power. */
+#if (defined(AIXPPC) || defined(LINUXPPC))
+				vm->enableGlobalLockReservation = 1;
+#endif
```
The outermost parentheses are not necessary here, but `#endif` lines should have a comment showing the matching `#if` expression.

IBMJimmyk replied:
I will remove the parentheses and add the comment. I'll also add similar comments to the `#endif` in `j9nonbuilder.h` and `ObjectMonitor.cpp` that I added.

gacholio commented:
I'm not liking to overcomplexity that's been happening lately - removing my approval.

IBMJimmyk commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5870,128 +5906,124 @@ TR::Register *J9::Power::TreeEvaluator::VMmonexitEvaluator(TR::Node *node, TR::C
 
-      generateLabelInstruction(cg, TR::InstOpCode::label, node, decLabel);
+      generateLabelInstruction(cg, TR::InstOpCode::label, node, decrementCheckLabel);
       if (cg->comp()->target().is64Bit())
```
I'll change this to check for compressed lockwords. I'll also change the one in VMmonentEvaluator that does a similar check before incrementing the RC field. The other 64bit checks in the lockword related evaluators seemed to be correct to me.

keithc-ca commented on jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:
```diff
@@ -1031,37 +1043,28 @@ public int getInitialLockword64(long j9clazz) {
 	 *          The initial lockword to use.
 	 */
-	public int getInitialLockword(int flags, long lockReservationHistory) {
+	public int getInitialLockword(int flags, long reservedCounter, long cancelCounter) {
```
Why are the counter parameters of type `long`? They're at most 16 bits - `int` would seem a better choice.

IBMJimmyk replied:
I'll change that. I was using `long` there before because `lockReservationHistory` was `long` sometimes but that shouldn't be needed anymore after splitting it up.

keithc-ca commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -3007,7 +3007,11 @@ typedef struct J9Class {
 	UDATA lockOffset;
-	UDATA lockReservationHistory;
+#if defined(J9VM_ENV_DATA64)
+	U_32 paddingForGLRCounters; // This is used to preserve alignment under 64 bit.
```
You can't use C++-style comments in this header file - it is also used by C code (and some compilers are pedantic).
Same goes for second `paddingForGLRCounters` declaration below.

IBMJimmyk replied:
Okay, I'll change those.

IBMJimmyk commented:
I've updated the branch with my latest changes. Things that still need to be worked on include potentially renaming ObjectMonitor_V1 to ObjectMonitor_V0 and ObjectMonitor_V2 to ObjectMonitor_V1. Do we actually want to do that? Also the issue related to changing existing defines in `j9nonbuilder.h`.

keithc-ca commented:
> renaming ObjectMonitor_V1 to ObjectMonitor_V0 and ObjectMonitor_V2 to ObjectMonitor_V1. Do we actually want to do that?

I don't think this is the time to begin renaming version-specific classes in DDR.

> Also the issue related to changing existing defines in `j9nonbuilder.h`.

Please clarify what you mean by that.

IBMJimmyk commented:
I'm referring to the previous comment you made about how some of the old values in `j9nonbuilder.h` are still needed for older corefiles. Is it only `OBJECT_HEADER_LOCK_RECURSION_OFFSET` that needs to keep its older version? I initially thought your comment meant that an older version of `OBJECT_HEADER_LOCK_RECURSION_MASK` needed to be kept as well but I think I was mistaken since it isn't used in `ObjectMonitor_V1.java`.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ObjectMonitor_V2.java:
```diff
@@ -0,0 +1,322 @@
+	@Override
+	public int hashCode() {
+		return super.hashCode();
+	}
```
Why not simply inherit this behavior?

IBMJimmyk replied:
That behavior was copied directly from `ObjectMonitor_V1.java` which does the exact same thing so I don't know the reasoning behind it. Should I get rid of it?

keithc-ca replied:
I don't think it helps, so I think it should be removed.

IBMJimmyk replied:
Okay, I'll remove it.

IBMJimmyk commented on runtime/oti/ObjectMonitor.hpp:
```diff
@@ -473,6 +481,115 @@ class VM_ObjectMonitor
+
+			clazz->reservedCounter = (reservedCounter >> 1) + 1;
+			clazz->cancelCounter = (cancelCounter >> 1);
+		}
```
I added a commit to remove `OBJECT_HEADER_LOCK_V2_FIRST_RECURSION_BIT` and `OBJECT_HEADER_LOCK_V2_RECURSION_MASK`.

Can you clarify what you are referring to by "the instance"?

There is no locking or anything to prevent concurrent updates to the class counters. So updates by one thread may get clobbered by updates from other threads. It was decided that this would be okay since the values of the counters don't need to keep perfect track of the reservation and cancellation events.

keithc-ca replied:
My point is that two threads might lock distinct objects (instances) of the same class. How do we not have a race with respect to classes?

zl-wang replied:
@keithc-ca we have to live with that race condition.  a conscious trade-off decision, that is. no correctness issue, but collected data driving the appropriate optimization.

keithc-ca replied:
I think that's something worth mentioning in `ObjectLockword.md`.

IBMJimmyk replied:
I've added a mention to it in the documentation.

IBMJimmyk commented on runtime/oti/monhelp.h:
```diff
@@ -26,11 +26,13 @@
  */
 #if defined(J9VM_THR_LOCK_RESERVATION)
-#define J9_FLATLOCK_COUNT(lock)  ((((lock) & OBJECT_HEADER_LOCK_BITS_MASK) >> OBJECT_HEADER_LOCK_RECURSION_OFFSET) + (((lock) & OBJECT_HEADER_LOCK_RESERVED) ? 0 : 1))
+#define J9_FLATLOCK_COUNT(lock)  ((((lock) & OBJECT_HEADER_LOCK_BITS_MASK) >> (((lock) & OBJECT_HEADER_LOCK_LEARNING) ? OBJECT_HEADER_LOCK_LEARNING_RECURSION_OFFSET : OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET)) + (((lock) & (OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_LEARNING)) ? 0 : 1))
```
That looks better to me. I'll change it.

IBMJimmyk commented on runtime/vm/jvminit.c:
```diff
@@ -7069,3 +7111,70 @@ setThreadNameAsyncHandler(J9VMThread *currentThread, IDATA handlerKey, void *use
+parseGlrOption(J9JavaVM* jvm, char* option)
+{
+	char* valueString = strstr(option, "=");
+	UDATA value;
```
Okay, I'll change that.

IBMJimmyk commented on runtime/vm/monhelpers.c:
```diff
@@ -54,35 +54,72 @@ objectMonitorExit(J9VMThread* vmStruct, j9object_t object)
+				casSuccess = (lock == compareAndSwapU32((uint32_t*)lockEA, (uint32_t)lock, (uint32_t)(lock - OBJECT_HEADER_LOCK_LEARNING_FIRST_RECURSION_BIT))) ? TRUE : FALSE;
+			} else {
+				casSuccess = (lock == compareAndSwapUDATA((uintptr_t*)lockEA, (uintptr_t)lock, (uintptr_t)(lock - OBJECT_HEADER_LOCK_LEARNING_FIRST_RECURSION_BIT))) ? TRUE : FALSE;
+			}
```
I'll get rid of the redundant `TRUE` and `FALSE` and split the line into two.

IBMJimmyk commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/ObjectMonitor_V2.java:
```diff
@@ -0,0 +1,322 @@
+					}
+					if(lockword.allBitsIn(OBJECT_HEADER_LOCK_LEARNING)) {
+						base.rightShift((int)OBJECT_HEADER_LOCK_LEARNING_RECURSION_OFFSET - (int)OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET);
+					}
```
I missed that update to `base`. I'll add it in and rearrange the if statement to follow yours.

keithc-ca commented on runtime/vm/ObjectMonitor.cpp:
```diff
@@ -283,31 +310,119 @@ objectMonitorEnterNonBlocking(J9VMThread *currentThread, j9object_t object)
 		} else {
 			/* check to see if object is unlocked (JIT did not do initial inline sequence due to insufficient static type info) */
-			if ((0 == (lock & ~OBJECT_HEADER_LOCK_RESERVED)) &&
+			if ((0 == (lock & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_RESERVED & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING)) &&
```
The extra cast obscures what is happening here:
```
if ((0 == (lock & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_RESERVED & ~(j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING)) &&
```
it could be rewritten as:
```
if ((0 == (lock & ~(j9objectmonitor_t)(OBJECT_HEADER_LOCK_RESERVED & OBJECT_HEADER_LOCK_LEARNING))) &&
```

IBMJimmyk replied:
I can change that. The check is supposed to check if every bit in the lockword except the reserved and learning bits are 0 so it will actually be `OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_LEARNING` before the cast.

keithc-ca replied:
Even better would be to make use of `J9_ARE_NO_BITS_SET` like elsewhere in this file.

IBMJimmyk replied:
That's a good idea. I just added that in.

keithc-ca commented on runtime/vm/ObjectMonitor.cpp:
```diff
@@ -283,31 +310,119 @@ objectMonitorEnterNonBlocking(J9VMThread *currentThread, j9object_t object)
 					/* lock is flat, am I the owner? */
 					if (J9_FLATLOCK_OWNER(lock) == currentThread) {
+						/* Convert Learning to Flat instead Inflated for dealing with RC overflow.*/
+						if (OBJECT_HEADER_LOCK_LEARNING == (lock & (OBJECT_HEADER_LOCK_LEARNING + OBJECT_HEADER_LOCK_INFLATED))) {
```
Bits should be combined with `|`, not `+`.

IBMJimmyk replied:
I'll change that.

keithc-ca commented on runtime/vm/ObjectMonitor.cpp:
```diff
@@ -283,31 +310,119 @@ objectMonitorEnterNonBlocking(J9VMThread *currentThread, j9object_t object)
+								 * For the same logical recursion count, the value encoded in the RC bits is one lower under
+								 * Flat compared to Learning so the new RC field is one lower than before the conversion to flat.
+								 */
+								newFlatLock = (lock & ~((j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING | (j9objectmonitor_t)OBJECT_HEADER_LOCK_RECURSION_MASK)) |
```
Merge casts; this could be:
```
newFlatLock = (lock & ~(j9objectmonitor_t)(OBJECT_HEADER_LOCK_LEARNING | OBJECT_HEADER_LOCK_RECURSION_MASK)) |
```

IBMJimmyk replied:
I'll change that.

keithc-ca commented on runtime/vm/ObjectMonitor.cpp:
```diff
@@ -406,9 +521,40 @@ spinOnFlatLock(J9VMThread *currentThread, j9objectmonitor_t volatile *lwEA, j9ob
+						 * For the same logical recursion count, the value encoded in the RC bits is one lower under
+						 * Flat compared to Learning so the new RC field is one lower than before the conversion to flat.
+						 */
+						j9objectmonitor_t const newLock = (lock & ~((j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING | (j9objectmonitor_t)OBJECT_HEADER_LOCK_RECURSION_MASK)) |
```
Merge casts:
```
j9objectmonitor_t const newLock = (lock & ~(j9objectmonitor_t)(OBJECT_HEADER_LOCK_LEARNING | OBJECT_HEADER_LOCK_RECURSION_MASK)) |

```

IBMJimmyk replied:
I changed that. I also found a few other places where the casts could be combined and also combined them.

gita-omr commented:
Hoping we can finish the full feature for 0.20:

- [ ] implemetation: incorporate all the suggestions

- [ ] performance testing

- [ ] setting up regular functional testing

- [ ] documentation

keithc-ca commented:
I think this is ready for PR testing. @IBMJimmyk could you please squash first?

IBMJimmyk commented:
Okay, I will squash the commits.

IBMJimmyk commented:
I also rebased on the latest master.

IBMJimmyk commented:
I think I might be seeing some unexpected performance issues in some tests I'm running. I am currently in the middle of investigating it.

IBMJimmyk commented:
Previously mentioned performance issue was a false alarm. It turned out to not be an actual problem.

I've made a few touchups to the documentation and rebased onto the latest master.

IBMJimmyk commented:
I just added a new commit that was fairly large. Please tell me if you feel like it might be better if it is reverted and added as a different PR.

The new commit moves the position of the Learning bit in the lockword. I found that it would be possible to reduce the JIT locking and unlocking path by 1 instruction if the RC field, Reserved bit and FLC bit weres consecutive bits in the lockword. This applies to 32 bit and 64bit with compressed lockwords. This faster path was added to `runtime/compiler/p/codegen/J9TreeEvaluator.cpp`.

Unfortunately, it turns out that `OBJECT_HEADER_LOCK_FLC` and `OBJECT_HEADER_LOCK_RESERVED` were used in `ObjectMonitor.java` so they can't be changed. As a result I had to add `OBJECT_HEADER_LOCK_V2_FLC` and `OBJECT_HEADER_LOCK_V2_RESERVED`. The bulk of the changes made were to add these V2 macros.

Documentation and comments were updated as well.

As for why I would like to add this into the current PR, it's so that I can avoid adding `V3` lockword macros for DDR corefile compatibility.

If you have time, please look at the new changes. @gacholio @keithc-ca

IBMJimmyk commented:
@zl-wang and @gita-omr might want to take a look as well.

IBMJimmyk commented:
I reconsidered adding the the previously mentioned commit since it was touching more places than I would like. I removed the commit and rebased onto latest master again to resolve a conflict.

IBMJimmyk commented:
Latest push is just minor documentation changes. There was incorrect capitalization in some of the options listed inside `ObjectLockword.md`. I fixed them.

IBMJimmyk commented:
Rebased onto latest master.

gita-omr commented:
Jenkins test sanity all jdk8,jdk11

gita-omr commented:
Jenkins test sanity plinux jdk8

gita-omr commented:
Restarted Build_JDK11_ppc64le_linux_Personal

IBMJimmyk commented:
Build_JDK11_ppc64le_linux_Personal failure:
https://ci.eclipse.org/openj9/job/Build_JDK11_ppc64le_linux_Personal/734/
```
[2020-03-06T02:07:40.491Z] IOException caught during compilation: Server failed to initialize: Resource deadlock avoided
[2020-03-06T02:07:40.491Z] gmake[3]: *** [/home/jenkins/workspace/Build_JDK11_ppc64le_linux_Personal/build/linux-ppc64le-normal-server-release/jdk/modules/java.desktop/_the.java.desktop_batch] Error 1
[2020-03-06T02:07:40.491Z] CompileJavaModules.gmk:590: recipe for target '/home/jenkins/workspace/Build_JDK11_ppc64le_linux_Personal/build/linux-ppc64le-normal-server-release/jdk/modules/java.desktop/_the.java.desktop_batch' failed
[2020-03-06T02:07:40.491Z] gmake[2]: *** [java.desktop-java] Error 1
[2020-03-06T02:07:40.491Z] make/Main.gmk:203: recipe for target 'java.desktop-java' failed
[2020-03-06T02:07:40.491Z] gmake[2]: *** Waiting for unfinished jobs....
```

I'm not entirely sure why that happened but I don't think it can be related to my change.

https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.functional_x86-64_linux_Personal/533/tapResults/
https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.functional_x86-64_linux_xl_Personal/277/tapResults/
https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_sanity.functional_x86-64_linux_Personal/565/tapResults/
https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_sanity.functional_x86-64_linux_xl_Personal/258/tapResults/

For all four of these there is 1 failing test which is testJITServer_1. It's a known issue that's unrelated to this PR: https://github.com/eclipse/openj9/issues/8770


The Travis CI build failed
https://travis-ci.org/eclipse/openj9/builds/658887189?utm_source=github_status&utm_medium=notification

This looks like it needs to be rerun with latest OMR. It was run when OpenJ9 and OMR were not compatible with each other.

gita-omr commented:
All failures are known and unrelated.

gita-omr commented:
Jenkins test extended all jdk8,jdk11

gita-omr commented:
All extended jobs passed. Merging. Thanks @IBMJimmyk  and everybody!
